src/openvic-simulation/misc/Modifier.cpp

```diff
@@ -110,8 +144,63 @@ bool TriggeredModifier::parse_scripts(DefinitionManager const& definition_manage
 	return trigger.parse_script(false, definition_manager);
 }
 
-ModifierInstance::ModifierInstance(Modifier const& modifier, Date expiry_date)
-	: modifier { modifier }, expiry_date { expiry_date } {}
+ModifierInstance::ModifierInstance(Modifier const& new_modifier, Date new_expiry_date)
+	: modifier { &new_modifier }, expiry_date { new_expiry_date } {}
+
+void ModifierSum::clear() {
+	modifiers.clear();
+	value_sum.clear();
+}
+
+bool ModifierSum::empty() {
+	return modifiers.empty();
+}
+
+fixed_point_t ModifierSum::get_effect(ModifierEffect const* effect, bool* successful) const {
+	return value_sum.get_effect(effect, successful);
```

Comment:
Why would get_effect not be successful?
Is this some sort of try catch for errors?

Reply:
It's to detect if the lookup is successful, i.e. if there is a value for the effect in the `ModifierSum` or `ModifierValue`. This allows us to tell the difference between a modifier with an effect explicitly listed with value 0 (e.g. tech schools include every research type modifier, even if the value is 0) vs a modifier having no entry for the effect at all.

Reply:
optional would be a better fit for that IMO instead of an out parameter (which you will then have to check for null)

Reply:
Null is allowed, it's the default value and just means we don't set the success flag. In most cases we don't care if the value is an explicit or implicit 0, and using optional would complicate those cases

Reply:
Could you rename it to something like is_present or has_value?
Successful sounds like a try catch.

Reply:
> Null is allowed, it's the default value and just means we don't set the success flag. In most cases we don't care if the value is an explicit or implicit 0, and using optional would complicate those cases

That is what the `value_or` member function is for ( https://en.cppreference.com/w/cpp/utility/optional/value_or )
If you don't care whether the value is present you would do `get_effect(...).value_or(0)` which is ergonomic and avoids easy-to-misuse out parameters